src/Common/GuardedPoolAllocator.cpp

```diff
@@ -0,0 +1,263 @@
+    return ptr & ~(pageSize - 1);
+}
+
+/*
+bool isPowerOfTwo(uintptr_t X) { return (X & (X - 1)) == 0; }
+*/
+} // anonymous namespace
+
+
+ALWAYS_INLINE GuardedPoolAllocator * getAllocator()
+{
+    return singletonPtr;
+}
+
+void *GuardedPoolAllocator::map(size_t size) const {
+    assert((size % state.pageSize) == 0);
+    void *ptr = mmap(nullptr, size, PROT_READ | PROT_WRITE,
+                    MAP_ANONYMOUS | MAP_PRIVATE, -1, 0);
+    check(ptr != MAP_FAILED, "Failed to map guarded pool allocator memory");
+    return ptr;
+}
+
+void *GuardedPoolAllocator::reserveGuardedPool(size_t size) {
+    assert((size % state.pageSize) == 0);
+    void *ptr =
+        mmap(nullptr, size, PROT_NONE, MAP_ANONYMOUS | MAP_PRIVATE, -1, 0);
+    check(ptr != MAP_FAILED, "Failed to reserve guarded pool allocator memory");
+    return ptr;
+}
+
+void GuardedPoolAllocator::allocateInGuardedPool(void *ptr, size_t size) const {
+    assert((reinterpret_cast<uintptr_t>(ptr) % state.pageSize) == 0);
+    assert((size % state.PageSize) == 0);
+    printf("Try to protect addr %p of size %zu\n", ptr, size);
+    fflush(stdout);
+    check(mprotect(ptr, size, PROT_READ | PROT_WRITE) == 0,
+        "Failed to allocate in guarded pool allocator memory");
+}
+
+void GuardedPoolAllocator::deallocateInGuardedPool(void *ptr,
+                                                   size_t size) const {
+    assert((reinterpret_cast<uintptr_t>(ptr) % state.pageSize) == 0);
+    assert((size % state.pageSize) == 0);
+    // mmap() a PROT_NONE page over the address to release it to the system, if
+    // we used mprotect() here the system would count pages in the quarantine
+    // against the RSS.
+    check(mmap(ptr, size, PROT_NONE, MAP_FIXED | MAP_ANONYMOUS | MAP_PRIVATE, -1,
+                0) != MAP_FAILED,
+        "Failed to deallocate in guarded pool allocator memory");
+}
+
+uintptr_t GuardedPoolAllocator::alignUp(uintptr_t ptr, size_t alignment) {
+    assert(isPowerOfTwo(alignment));
+    assert(alignment != 0);
+    if ((ptr & (alignment - 1)) == 0)
+        return ptr;
+
+    printf("Align up; add %lu to pointer\n", alignment - (ptr & (alignment - 1)));
+    printf("Old=%p; new=%p\n", reinterpret_cast<void *>(ptr), reinterpret_cast<void *>(ptr + alignment - (ptr & (alignment - 1))));
+    ptr += alignment - (ptr & (alignment - 1));
+    return ptr;
+}
+
+uintptr_t GuardedPoolAllocator::alignDown(uintptr_t ptr, size_t alignment) {
+    assert(isPowerOfTwo(alignment));
+    assert(alignment != 0);
+    if ((ptr & (alignment - 1)) == 0)
+        return ptr;
+
+    printf("Old=%p; new=%p\n", reinterpret_cast<void *>(ptr), reinterpret_cast<void *>(ptr - (ptr & (alignment - 1))));
+    printf("Align down; subtract %lu from pointer\n", ptr & (alignment - 1));
+    ptr -= ptr & (alignment - 1);
+    return ptr;
+}
+
+void GuardedPoolAllocator::init()
+{
+    printf("Initialize gwp asan\n");
+    singletonPtr = this;
```

Comment:
This should be moved to the end of the function. Also it is racy now, right?

Reply:
As I see it, `init` will be called before other static initializators [here](https://github.com/ClickHouse/ClickHouse/pull/36826/files#diff-e218677ebcdab2da6b73a676c14ad9489ad3e2b9360e1ac039d24f55c7b51740R351).
Therefore it will be run even before entering main, therefore before spawning any other threads. So I have come to a conclusion that it will be thread-safe, however, I am open to other opinions on this since I am not particularly knowledgeable in this area :)

As for moving it to the end of the function – will do it in the next update.